MoltenVK/MoltenVK/GPUObjects/MVKDescriptor.mm

```diff
@@ -490,19 +485,50 @@ void mvkPopulateShaderConversionConfig(mvk::SPIRVToMSLConversionConfiguration& s
 	[args addObject: argDesc];
 }
 
+uint32_t MVKDescriptorSetLayoutBinding::getMTLResourceCountPerElement() {
+	switch (_info.descriptorType) {
+		case VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER:
+			return 2;
+		case VK_DESCRIPTOR_TYPE_STORAGE_IMAGE:
+		case VK_DESCRIPTOR_TYPE_STORAGE_TEXEL_BUFFER:
+			return getMetalFeatures().nativeTextureAtomics ? 1 : 2;
+		default:
+			return 1;
+	}
+}
+
+uint64_t MVKDescriptorSetLayoutBinding::getMetal3ArgumentBufferEncodedSize() {
+	return getMTLResourceCountPerElement() * getDescriptorCount() * kMVKMetal3ArgBuffSlotSizeInBytes;
+}
```

Comment:
Multi-plane tracking is problematic, because there does not seem to be any way to know at descriptor set allocation time, that any descriptor might be assigned a multi-plane image, so we'll have to reserve 3 texture slots for all images.

This also affects discrete texture indexing as well, since any of those could be a multi-plane image as well.

So far, we seem to be getting away with this simply because CTS only tests with one image per descriptor set.

So far, I think the best we can do is that when `samplerYcbcrConversion` or `VK_KHR_sampler_ycbcr_conversion` is enabled, allocate 3 argument buffer slots per image descriptor, and otherwise (the more common use case) allocate 1 argument buffer slot per image descriptor.

Anyway, since this is messy, and is not strictly tied to argument buffers, I decided to address this in a separate PR, which I'll start to work on once this PR is finished.

Longer term, another option might be to add a Vulkan extension to add a flag to `VkDescriptorBindingFlagBits` to indicate a multi-plane image might be stored in that descriptor (or maybe the reverse, that it definitely won't).

I'd love to hear any thoughts you might have about handling this.

Reply:
> So far, I think the best we can do is that when `samplerYcbcrConversion` or `VK_KHR_sampler_ycbcr_conversion` is enabled, allocate 3 argument buffer slots per image descriptor, and otherwise (the more common use case) allocate 1 argument buffer slot per image descriptor.

Yeah, that's what I was thinking.

> Anyway, since this is messy, and is not strictly tied to argument buffers, I decided to address this in a separate PR, which I'll start to work on once this PR is finished.

Very well.

> Longer term, another option might be to add a Vulkan extension to add a flag to `VkDescriptorBindingFlagBits` to indicate a multi-plane image might be stored in that descriptor (or maybe the reverse, that it definitely won't).

That actually sounds like something that should've been part of `VK_KHR_sampler_ycbcr_conversion` in the first place. I wonder how other drivers handle this situation...

As you probably know, I can't push for this in Khronos, since I'm no longer a member.